xapian-letor/feature/tfdoclencolltfcolllenfeature.cc

```diff
@@ -53,94 +55,65 @@ TfDoclenCollTfCollLenFeature::get_values() const
 
     vector<double> values;
     double value = 0;
-    double coll_len;
-    double doc_len;
-    auto coll_len_iterator = collection_length.find("title");
-    if (coll_len_iterator != collection_length.end())
-	coll_len = (double)coll_len_iterator->second;
-    else
+    double coll_len = 0;
+    if (!internal->get_collection_length("title", coll_len)) {
 	coll_len = 0;
-    auto doc_len_iterator = doc_length.find("title");
-    if (doc_len_iterator != doc_length.end())
-	doc_len = (double)doc_len_iterator->second;
-    else
+    }
+    double doc_len = 0;
+    if (!internal->get_doc_length("title", doc_len)) {
 	doc_len = 0;
+    }
 
+    Xapian::Query feature_query = internal->get_query();
     for (TermIterator qt = feature_query.get_unique_terms_begin();
 	 qt != feature_query.get_terms_end(); ++qt) {
 	if (is_title_term((*qt))) {
-	    double tf;
-	    double coll_tf;
-	    auto tf_iterator = termfreq.find(*qt);
-	    auto coll_tf_iterator = collection_termfreq.find(*qt);
-	    if (tf_iterator != termfreq.end())
-		tf = (double)tf_iterator->second;
-	    else
+	    double tf = 0;
+	    double coll_tf = 0;
+	    if (!internal->get_termfreq(*qt, tf))
 		tf = 0;
-	    if (coll_tf_iterator != collection_termfreq.end())
```

Comment:
The pattern "return `bool` indicating success, and the actual value found by reference" makes for somewhat clumsy code here.  It's also what's presumably led you to an API where everything's returned as a `double` regardless of what type is actually is.

Under what circumstances would we actually not find an entry when looking up in these maps?

If it should never happen, then I don't think it makes sense to complicate the API here to allow for it - we can either `Assert(false);` and `return 0;` if it's an error in internal state, or if it can be caused by bad parameters passed from user code throw `Xapian::InvalidArgumentError`.

If it can happen, but there's a sensible default (it looks like we actually always just use zero if the term is not found) then we can just make `get_termfreq()` etc return that default if they don't find the answer.

Reply:
> Under what circumstances would we actually not find an entry when looking up in these maps?
> 
It can happen when a feature query contains terms which are not in the feature database.
I am going with the latter option you suggested. Correct me if I am wrong.

Reply:
I guess it depends what we do for a term which doesn't exist when we fill these maps (I assume we must fill them based on terms from the query - creating them for all terms in the database would be a lot of unnecessary work).

It probably makes more sense not to bother inserting zero (or other default) values into the map for such terms, so defaulting on lookup if the term isn't found sounds good.